### PR TITLE
[SYCL-MLIR]: FunctionType and Attributes Infrastructure

### PR DESCRIPTION
**Pull request by etiotto**
In order to abide by the ABI calling conventions clang uses an exhaustive infrastructure. In cgeist we take a number of shortcuts. This is problematic because:
  - the cgeist implementation to add attributes to function types, parameters, call arguments is incomplete
  - the current way cgeist computes function attributes does not take into account the ABI constraints 
  
This PR enhances the way cgeist attaches function attributes to a function definition.

### CODE REVIEW

File: polygeist/tools/cgeist/Lib/Attributes.cc [hunk 1]
```diff
@@ -40,13 +40,28 @@ static void addToPassThroughAttr(NamedAttribute &PassThroughAttr,
...
+    if (auto StrA1 = A1.dyn_cast<StringAttr>()) {
+      if (auto StrA2 = A2.dyn_cast<StringAttr>())
+        return StrA1 < StrA2;
+      return true;
+    }
+    if (auto ArrA1 = A1.dyn_cast<ArrayAttr>()) {
+      if (auto ArrA2 = A2.dyn_cast<ArrayAttr>())
+        return ArrA1[0].cast<StringAttr>() < ArrA2[0].cast<StringAttr>();
+      return false;
+    }
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Lib/Attributes.cc:61]: ` if (auto ArrA1 = A1.dyn_cast<ArrayAttr>()) {` will always be true. Can we replace with `cast` and no `if`?

File: polygeist/tools/cgeist/Lib/Attributes.cc [hunk 2]
```diff
@@ -40,13 +40,28 @@ static void addToPassThroughAttr(NamedAttribute &PassThroughAttr,
   std::vector<mlir::Attribute> Vec =
       PassThroughAttr.getValue().cast<ArrayAttr>().getValue().vec();
 
-  // TODO: find a way to add the attributes only if one does not exist already,
-  // and keep the list in sorted order.
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Lib/Attributes.cc:44]: We've fixed the first condition in the comment, but not the second, right?

Comment (victor-eds) [reply polygeist/tools/cgeist/Lib/Attributes.cc:44]: No need to do now, but could keep comment (relevant part)

File: polygeist/tools/cgeist/Test/Verification/sycl/kernels.cpp [hunk 3]
```diff
@@ -82,4 +81,4 @@ SYCL_EXTERNAL void function_1(sycl::item<2, true> item) {
 }
 
 // Keep at the end of the file.
-// CHECK-LLVM: attributes #1 = { convergent mustprogress norecurse nounwind "sycl-module-id"="{{.*}}/polygeist/tools/cgeist/Test/Verification/sycl/kernels.cpp" }
+// CHECK-LLVM: attributes #1 = { convergent mustprogress norecurse nounwind "frame-pointer"="all" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "sycl-module-id"="{{.*}}/polygeist/tools/cgeist/Test/Verification/sycl/kernels.cpp" }
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Test/Verification/sycl/kernels.cpp:84]: No new line at the end of file

**Review by victor-eds — COMMENTED** [review]
LGTM but for a couple of comments.

Also, dropping MLIR function attributes from tests makes sense, as we will be testing those through LLVM translation, but I'd argue we might want to keep those given the incomplete LLVM IR to LLVM attributes conversion. What do you think?

File: polygeist/tools/cgeist/Lib/clang-mlir.cc [hunk 4]
```diff
@@ -3469,25 +3489,250 @@ void MLIRASTConsumer::setMLIRFunctionVisibility(
...
+}
+
+void MLIRASTConsumer::setMLIRFunctionAttributesForDefinition(
+    const Decl *D, mlir::FunctionOpInterface F) const {
+  const CodeGenOptions &CodeGenOpts = CGM.getCodeGenOpts();
+  const LangOptions &LangOpts = CGM.getLangOpts();
+  MLIRContext *Ctx = module->getContext();
+  mlirclang::AttrBuilder B(*Ctx);
+
+  if (CodeGenOpts.UnwindTables)
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/clang-mlir.cc:3520]: `SetLLVMFunctionAttributesForDefinition` has the following:
```suggestion
  if ((!D || !D->hasAttr<NoUwtableAttr>()) && CodeGenOpts.UnwindTables)
```

File: polygeist/tools/cgeist/Lib/clang-mlir.cc [hunk 5]
```diff
@@ -3469,25 +3489,250 @@ void MLIRASTConsumer::setMLIRFunctionVisibility(
...
+
+  if (CodeGenOpts.StackClashProtector)
+    B.addPassThroughAttribute("probe-stack",
+                              mlir::StringAttr::get(Ctx, "inline-asm"));
+
+  if (!hasUnwindExceptions(LangOpts))
+    B.addPassThroughAttribute(llvm::Attribute::NoUnwind);
+
+  if (!D || !D->hasAttr<NoStackProtectorAttr>()) {
+    if (LangOpts.getStackProtector() == LangOptions::SSPOn)
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/clang-mlir.cc:3533]: Missing this case?
```
if (D && D->hasAttr<StrictGuardStackCheckAttr>() &&
           LangOpts.getStackProtector() == LangOptions::SSPOn)
    B.addAttribute(llvm::Attribute::StackProtectStrong);
```

File: polygeist/tools/cgeist/Lib/Attributes.cc (hunk 2 above)
Comment (etiotto) [reply polygeist/tools/cgeist/Lib/Attributes.cc:44]: Correct I can add a comment.

File: polygeist/tools/cgeist/Lib/Attributes.cc (hunk 1 above)
Comment (etiotto) [reply polygeist/tools/cgeist/Lib/Attributes.cc]: done

File: polygeist/tools/cgeist/Lib/CGCall.cc [hunk 6]
```diff
@@ -1641,7 +1641,8 @@ ValueCategory MLIRScanner::VisitCallExpr(clang::CallExpr *expr) {
     ShouldEmit = true;
 
   FunctionToEmit F(*callee, mlirclang::getInputContext(builder));
-  auto ToCall = cast<func::FuncOp>(Glob.GetOrCreateMLIRFunction(F, ShouldEmit));
+  auto ToCall = cast<func::FuncOp>(
+      Glob.GetOrCreateMLIRFunction(F, false /*IsThunk */, ShouldEmit));
```
[the remark this answers is not in the document]
Comment (etiotto) [reply polygeist/tools/cgeist/Lib/CGCall.cc]: OK I have removed it for now. We can add it back when we find the need.

File: polygeist/tools/cgeist/Lib/clang-mlir.cc (hunk 5 above)
Comment (etiotto) [reply polygeist/tools/cgeist/Lib/clang-mlir.cc]: Thanks, added.

File: polygeist/tools/cgeist/Lib/clang-mlir.cc [hunk 7]
```diff
@@ -3469,25 +3489,250 @@ void MLIRASTConsumer::setMLIRFunctionVisibility(
...
+#endif
+}
+
 void MLIRASTConsumer::setMLIRFunctionAttributes(
     mlir::FunctionOpInterface function, const FunctionToEmit &FTE,
-    bool ShouldEmit) {
+    const clang::CodeGen::CGFunctionInfo &FI, bool IsThunk, bool ShouldEmit) {
   using Attribute = llvm::Attribute;
 
   const FunctionDecl &FD = FTE.getDecl();
```
[the remark this answers is not in the document]
Comment (etiotto) [reply polygeist/tools/cgeist/Lib/clang-mlir.cc]: done

File: polygeist/tools/cgeist/Test/Verification/sycl/kernels.cpp (hunk 3 above)
Comment (etiotto) [reply polygeist/tools/cgeist/Test/Verification/sycl/kernels.cpp]: added

File: polygeist/tools/cgeist/Lib/clang-mlir.cc (hunk 4 above)
Comment (etiotto) [reply polygeist/tools/cgeist/Lib/clang-mlir.cc:3519]: Must be something that has been updated after we branched. Added.

**Comment by etiotto** [comment]
> LGTM but for a couple of comments.
> 
> Also, dropping MLIR function attributes from tests makes sense, as we will be testing those through LLVM translation, but I'd argue we might want to keep those given the incomplete LLVM IR to LLVM attributes conversion. What do you think?

I have kept some of the checks in MLIR for 'passthrough' attributes in some tests to ensure we can test that aspect. I have removed them in most files given that we care about the attributes being lowered correctly to LLVM IR and we have tests for that.

**Review by whitneywhtsang — APPROVED** [review]

**Pull request merged**